src/lib.rs

```diff
@@ -24,89 +22,138 @@ pub mod prelude;
-/// register_module!(mut cx, {
-///     cx.export_function("foo", foo)?;
-///     cx.export_function("bar", bar)?;
-///     cx.export_function("baz", baz)?;
-///     Ok(())
-/// });
-/// ```
-#[macro_export]
-macro_rules! register_module {
-    ($module:pat, $init:block) => {
-        // Mark this function as a global constructor (like C++).
-        #[allow(improper_ctypes)]
-        #[cfg_attr(target_os = "linux", link_section = ".ctors")]
-        #[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
-        #[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
-        #[used]
-        pub static __LOAD_NEON_MODULE: extern "C" fn() = {
-            fn __init_neon_module($module: $crate::context::ModuleContext) -> $crate::result::NeonResult<()> $init
-
-            extern "C" fn __load_neon_module() {
-                // Put everything else in the ctor fn so the user fn can't see it.
-                #[repr(C)]
-                struct __NodeModule {
-                    version: i32,
-                    flags: u32,
-                    dso_handle: *mut u8,
-                    filename: *const u8,
-                    register_func: Option<extern "C" fn(
-                        $crate::handle::Handle<$crate::types::JsObject>, *mut u8, *mut u8)>,
-                    context_register_func: Option<extern "C" fn(
-                        $crate::handle::Handle<$crate::types::JsObject>, *mut u8, *mut u8, *mut u8)>,
-                    modname: *const u8,
-                    priv_data: *mut u8,
-                    link: *mut __NodeModule
+#[cfg(all(feature = "legacy-runtime", feature = "napi-runtime"))]
+compile_error!("Cannot enable both `legacy-runtime` and `napi-runtime` features.\n\nTo use `napi-runtime`, disable `legacy-runtime` by setting `default-features` to `false` in Cargo.toml\nor with cargo's --no-default-features flag.");
+
+use cfg_if::cfg_if;
+
+cfg_if! {
+    if #[cfg(feature = "napi-runtime")] {
+        /// Register the current crate as a Node module, providing startup
+        /// logic for initializing the module object at runtime.
+        ///
+        /// The first argument is a pattern bound to a `neon::context::ModuleContext`. This
+        /// is usually bound to a mutable variable `mut cx`, which can then be used to
+        /// pass to Neon APIs that require mutable access to an execution context.
+        ///
+        /// Example:
+        ///
+        /// ```rust,ignore
+        /// register_module!(mut cx, {
+        ///     cx.export_function("foo", foo)?;
+        ///     cx.export_function("bar", bar)?;
+        ///     cx.export_function("baz", baz)?;
+        ///     Ok(())
+        /// });
+        /// ```
+        #[macro_export]
+        macro_rules! register_module {
+            ($module:pat, $init:block) => {
+                register_module!(env!("CARGO_PKG_NAME"), $module, $init);
+            };
+
+            ($name:expr, $module:pat, $init:block) => {
```

Comment:
I don't expect `$name` to be used. Is it being kept to minimize breaking changes?

Reply:
We don't want to expose the N-API data structures to the programmer. The lack of hygiene is unfortunate but I wouldn't want to pollute the UI with implementation details.

Re `$name`, I think I copied that from your PoC 😛 We could separate that out into a doc-hidden helper macro.

Reply:
Oh ha, `$name` is unused now! Cleanup coming…